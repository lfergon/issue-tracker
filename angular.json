{
  "$schema": "./node_modules/@angular/cli/lib/config/schema.json",
  "version": 1,
  "newProjectRoot": "projects",
  "projects": {
    "issue-tracker": {
      "projectType": "application",
      "schematics": {},
      "root": "",
      "sourceRoot": "src",
      "prefix": "app",
      "architect": {
        "build": {
          "builder": "@angular-devkit/build-angular:browser",
          "options": {
            "outputPath": "dist/issue-tracker",
            "index": "src/index.html",
            "main": "src/main.ts",
            "polyfills": ["zone.js"],
            "tsConfig": "tsconfig.app.json",
            "assets": ["src/favicon.ico", "src/assets"],
            "styles": ["node_modules/@clr/ui/clr-ui.min.css", "src/styles.css"],
            "scripts": []
          },
          "configurations": {
            "production": {
              "budgets": [
                {
                  "type": "initial",
                  "maximumWarning": "500kb",
                  "maximumError": "1mb"
                },
                {
                  "type": "anyComponentStyle",
                  "maximumWarning": "2kb",
                  "maximumError": "4kb"
                }
              ],
              "outputHashing": "all"
            },
            "development": {
              "buildOptimizer": false,
              "optimization": false,
              "vendorChunk": true,
              "extractLicenses": false,
              "sourceMap": true,
              "namedChunks": true
            }
          },
          "defaultConfiguration": "production"
        },
        "serve": {
          "builder": "@angular-devkit/build-angular:dev-server",
          "configurations": {
            "production": {
              "buildTarget": "issue-tracker:build:production"
            },
            "development": {
              "buildTarget": "issue-tracker:build:development"
            }
          },
          "defaultConfiguration": "development"
        },
        "extract-i18n": {
          "builder": "@angular-devkit/build-angular:extract-i18n",
          "options": {
            "buildTarget": "issue-tracker:build"
          }
        },
        "test": {
          "builder": "@angular-devkit/build-angular:karma",
          "options": {
            "polyfills": ["zone.js", "zone.js/testing"],
            "tsConfig": "tsconfig.spec.json",
            "assets": ["src/favicon.ico", "src/assets"],
            "styles": ["src/styles.css"],
            "scripts": []
          }
        },
        "lint": {
          "builder": "@angular-eslint/builder:lint",
          "options": {
            "lintFilePatterns": ["src/**/*.ts", "src/**/*.html"]
          }
        }
      }
    }
  },
  "cli": {
    "analytics": "7b562994-7e0c-4e60-a27b-26d736495c7e",
    "schematicCollections": ["@angular-eslint/schematics"]
  }
}
