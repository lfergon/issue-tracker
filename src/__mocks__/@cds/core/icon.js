module.exports = {
  ClarityIcons: {
    addIcons: jest.fn()
  }
};