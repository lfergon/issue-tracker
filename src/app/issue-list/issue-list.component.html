@if (!showReportIssue) {
  <button class="btn btn-primary" (click)="showReportIssue = true">
    Add new issue
  </button>
  <clr-datagrid>
    <clr-dg-column [clrDgField]="'issueNumber'" [clrDgColType]="'number'"
      >Issue No</clr-dg-column
    >
    <clr-dg-column [clrDgField]="'type'">Type</clr-dg-column>
    <clr-dg-column [clrDgField]="'title'">Title</clr-dg-column>
    <clr-dg-column [clrDgField]="'description'">Description</clr-dg-column>
    <clr-dg-column [clrDgField]="'priority'">Priority</clr-dg-column>
    <clr-dg-row *clrDgItems="let issue of issues">
      <clr-dg-action-overflow>
        <button class="action-item" (click)="setSelectedIssue(issue)">
          Resolve
        </button>
      </clr-dg-action-overflow>
      <clr-dg-cell>{{ issue.issueNumber }}</clr-dg-cell>
      <clr-dg-cell>{{ issue.type }}</clr-dg-cell>
      <clr-dg-cell>{{ issue.title }}</clr-dg-cell>
      <clr-dg-cell>{{ issue.description }}</clr-dg-cell>
      <clr-dg-cell>
        <span class="label" [class.label-danger]="issue.priority === 'high'">{{
          issue.priority
        }}</span>
      </clr-dg-cell>
    </clr-dg-row>
    <clr-dg-footer>{{ issues.length }} issues</clr-dg-footer>
  </clr-datagrid>
}

@if (showReportIssue) {
  <app-issue-report (formClose)="onCloseReport()"></app-issue-report>
}

@if (selectedIssue) {
  <app-confirm-dialog
    [issueNumber]="selectedIssue.issueNumber"
    (confirm)="onConfirm($event)"
  ></app-confirm-dialog>
}
