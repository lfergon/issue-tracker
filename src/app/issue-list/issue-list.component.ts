import { Component, OnInit } from '@angular/core';
import { IssuesService } from '../issues.service';
import { Issue } from '../issue';
import { issues } from '../../assets/mock-issues';

@Component({
  selector: 'app-issue-list',
  templateUrl: './issue-list.component.html',
  styleUrls: ['./issue-list.component.css'],
})
export class IssueListComponent implements OnInit {
  showReportIssue = false;

  selectedIssue: Issue | undefined = undefined;

  protected readonly issues = issues;

  constructor(private issuesService: IssuesService) {}

  private getIssues() {
    return this.issuesService.getPendingIssues();
  }

  ngOnInit(): void {
    this.getIssues();
  }

  onCloseReport() {
    this.showReportIssue = false;
    this.getIssues();
  }

  onConfirm(confirmed: boolean) {
    if (confirmed && this.selectedIssue) {
      this.issuesService.completeIssue(this.selectedIssue);
      this.getIssues();
    }
    this.selectedIssue = undefined;
  }
}
