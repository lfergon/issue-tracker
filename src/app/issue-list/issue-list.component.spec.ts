import { ComponentFixture, TestBed } from '@angular/core/testing';

import { IssueListComponent } from './issue-list.component';
import { IssuesService } from '../issues.service';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { Mock } from '@testing-library/angular/jest-utils';

describe('IssueListComponent', () => {
  let component: IssueListComponent;
  let fixture: ComponentFixture<IssueListComponent>;
  let mockIssuesService: IssuesService;

  beforeEach(async () => {
    mockIssuesService = {
      getPendingIssues: jest.fn(),
      completeIssue: jest.fn(),
    };

    TestBed.configureTestingModule({
      declarations: [IssueListComponent],
      providers: [{ provide: IssuesService, useValue: mockIssuesService }],
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
    });

    fixture = TestBed.createComponent(IssueListComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should call getPendingIssues on init', () => {
    expect(mockIssuesService.getPendingIssues).toHaveBeenCalled();
  });

  it('should call completeIssue on onConfirm', () => {
    component.selectedIssue = {
      issueNumber: 5,
      title: '',
      description: '',
      priority: 'low',
      type: 'bug',
    };
    component.onConfirm(true);
    expect(mockIssuesService.completeIssue).toHaveBeenCalled();
  });

  it('should not call completeIssue on onConfirm if not confirmed', () => {
    component.selectedIssue = {
      issueNumber: 4,
      title: '',
      description: '',
      priority: 'medium',
      type: 'feature',
    };
    component.onConfirm(false);
    expect(mockIssuesService.completeIssue).not.toHaveBeenCalled();
    expect(component.selectedIssue).toBeUndefined();
  });
});
