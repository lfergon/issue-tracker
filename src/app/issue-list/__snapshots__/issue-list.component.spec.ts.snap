// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssueListComponent should match snapshot 1`] = `
<app-issue-list
  issues={[Function Array]}
  issuesService={[Function Object]}
  selectedIssue="undefined"
  showReportIssue="false"
>
  <button
    class="btn btn-primary"
  >
    Add new issue
  </button><clr-datagrid>
    <clr-dg-column>
      Issue No
    </clr-dg-column>
    <clr-dg-column>
      Type
    </clr-dg-column>
    <clr-dg-column>
      Title
    </clr-dg-column>
    <clr-dg-column>
      Description
    </clr-dg-column>
    <clr-dg-column>
      Priority
    </clr-dg-column>
    <clr-dg-footer>
      5 issues
    </clr-dg-footer>
  </clr-datagrid>
</app-issue-list>
`;
